### DISCUSSION TITLE
No TTY, No CP-A

### DISCUSSION THREAD

**Discussion by stamatovski** [thehighnibble/thehighnibble.github.io#177] [IMSAI8080esp replica]
The esp boots and starts the AP. I log in and I see the message NO CPA in the web ui, while i can see the web cpa and it works also showing green. Pressing on the buttons on the web interface changes the state like start and stop on the physical panel.

I also cant see anything on the TTY.
I have flashed 1.13 firmware and made new sdcard with the image. I used the web js flasher from espressiff and set the addresses for the bins same to what esptool uses in console. When doing it from scratch the instructions point to "just unzip the contents" which unzips IMAGE folder. You need to copy whats in the IMAGE folder to the root of the sdcard.

**Comment by nbreeden2** [comment]
The IMAGE zip gets unzipped at the root of the SD card. There are, I believe, 3 folders then at the root, IMSAI, CROMEMCO and maybe SYSTEM. There should be 3 root folders.

Are you able to use the Web UI? See the disk images in the Library? Look at the system configuration etc? See the virtual CP-A?

I've never tried to flash the firmware using the "web js flasher" nor have I ever had to "set the addresses" manually. I've flashed many times from the command line (pytool.py ?), using the drag&drop and the OTA methods Dave supports.

Have you set the NVS configuration on the system? It needs to be manually setup. One of the configurations allows you to boot CP/M.
https://thehighnibble.com/imsai8080/configure/#startup-configuration-non-volatile-storage-nvs

To boot CP/M you need to drag the CP/M floppy disk image from the Library to the A: drive in the Web UI (I don't believe it's mounted by default). You then need to reset/run the system via the CP-A.

**Comment by stamatovski** [comment]
This is a reminder for myself. RTFM Igor!!!, then watch Daves videos from
multiple places like youtube and the high nibble site and elsewhere...... I
had broken microsd card, so i had to do image on a brand new one. In the
process i have used imaging the esp from the web and that works without
using the python tool, quite nifty if you dont have python installed, and
the imaging works. The step that i forgot to do is setting the nvram. Here
are some links if anyone else finds it useful.
https://espressif.github.io/esptool-js/
However, the setup should at least include one working configuration from
scratch, people having those units for more than a month will forget that
you need to set up roms for booting up, not to mention a brand new happy
owner of this wonderful device.

Fixed, all working now :)
Regards

**Reply by nbreeden2** [reply]
Glad you got it working. :)

So, back in the day when you first powered up your new IMSAI kit you had some LEDs and switches. The system didn't even do a power on reset. It might wake up running or wake up halted. There was no OS, there was no BIOS, or ROM. There was a very simple test program in the assemble manual to do a quick test to see if it works. What's you've just experienced is that joy/frustration all over again which, in the end is what Dave's kit is all about.
You can reset the NVRAM to RAM only, the 8080 at 2MHz and step back into time to know what it was like in 1976 to have an IMSAI on the tabletop. Everything a user did back then took reading over the sparse documentation, reading the 8080 databook and programmers guide if they were lucky enough to have one. Issues could take weeks or months to resolve as you searched the few magazines and newsletters, and, if you were lucky a local computer user group or someone who had a working system. If you wanted to add a serial or parallel I/O card you had to read the docs, figure out the ports to use and how to strap the card to support the selected ports. If you wanted to add a floppy controller there was a real uphill battle. You have the controller card, a floppy drive and some manuals. If there wasn't a configuration of CP/M built for your hardware configuration you needed to create it, yet you couldn't without a bootable system. The CP/M manuals had a section on how to modify CP/M for your hardware.
In the modern world there is an expectation of plug it in, turn it on and use it. We're spoiled.
Back in the day the expectation was, plug it in, turn it on then figure out how to do something with it. For me this was/is the sweet spot. You need to dive into the hardware, learn assembly language, learn HEX, understand the 8080 and what the LEDs mean in the context of the 8080, learn to use the front panel and when eventually you got an LED to blink it felt like a real victory as you had to learn everything to get to this point.
Today you buy an Arduino, load a development tool on a PC, drop in some source code, compile it and the LED blinks. What, however, have you actually learned?
Embrace the struggle, enjoy the frustration. Work to understand what the "Programmed Input" switches do, how to read them and what the "Programmed Output" LEDs do and how to write to them. You can completely explore this and learn how to control them using only the front panel, a sheet of paper, a pencil and an 8080 assembler book. This opens the door to truly understanding the IMSAI hardware and its design.
When you can tell me why writing a zero to a programmed output LED turns it on while a 1 turns it off (seems backwards) and why this is a consequence of how the hardware is designed you'll be, IMHO, on the right journey. The answers to this are in the original IMSAI manual and schematics.

Again, glad you got it working. :)

**Comment by stamatovski** [comment]
I totally agree. On all points that you mentioned. The motive for me
sounding fast, is that I took off the unit from an ongoing exhibition in
Stockholm which lasts till 8th of november which gives me a week or so,
where it proudly sits among other authentic machines as a representative of
its era. Switches and blinking leds make it the most visually interesting
piece up there. The most asked question that people go for is "how do you
program imsai" which usually translates to what the switches do. Many find
it amazing to have the run stop and next step exposed on a switch,
comparing it to todays on screen debuggers of their flavor programming
language.
IMHO i think the imsai is the easiest and most visual entry into learning
assembly, where one actually sees the bits, learns about addresses, values
in binary and how they relate to hex, and reveals that a number can be a
cpu instruction. Then interrupts, which beg for temporary storage of
registers of your running program, stack heap etc. That is true
for altairs and pdps of different sorts. Much of that knowledge helps
revive old machines too. Like making a nop socket for 6502 where you feed
the cpu the same nop instruction over and over and it just loops through
the address bus, quite useful for testing the address bus and attached TTLs.
At the time of imsai, the line between a coder which makes a function and a
coder which uses that function was not so far apart. Today we have 3 lines
of code like model.load(), model.train(), model.run() and it is usually
accompanied by a shout "look ma i am an ai programmer", while the arduino
example looks super complex.
I am glad that I have the machine in a working state, it will be back on
display, glad that there are still people with joy and passion for
the bransche, thank you for the reply Neil!
